### ISSUE TITLE
map_unwrap_or doesn't trigger on references

### ISSUE THREAD

**Issue by TomFryersMidsummer** [rust-lang/rust-clippy#15713]
### Summary

This code:
```rust
#[deny(clippy::map_unwrap_or)]
fn main() {
    let x = Some(3);
    println!("{}", x.map(|y| y + 1).unwrap_or(3));
}
```
produces the following warning
```
$ clippy-driver main.rs
error: called `map(<f>).unwrap_or(<a>)` on an `Option` value
 --> main.rs:4:20
  |
4 |     println!("{}", x.map(|y| y + 1).unwrap_or(3));
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_unwrap_or
note: the lint level is defined here
 --> main.rs:1:8
  |
1 | #[deny(clippy::map_unwrap_or)]
  |        ^^^^^^^^^^^^^^^^^^^^^
help: use `map_or(<a>, <f>)` instead
  |
4 -     println!("{}", x.map(|y| y + 1).unwrap_or(3));
4 +     println!("{}", x.map_or(3, |y| y + 1));
  |

error: aborting due to 1 previous error
```

Which should apply equally when `x = &Some(3)`. It also misses references to `Results` for the `Result::unwrap_or_else` case.

### Lint Name

map_unwrap_or

### Reproducer

I tried this code:

```rust
#[deny(clippy::map_unwrap_or)]
fn main() {
    let x = &Some(3);
    println!("{}", x.map(|y| y + 1).unwrap_or(3));
}
```
I expected to see this happen: same or similar warning to above.

Instead, this happened: no warning.

### Version

```text
rustc 1.90.0 (1159e78c4 2025-09-14)
binary: rustc
commit-hash: 1159e78c4747b02ef996e55082b704c09b970588
commit-date: 2025-09-14
host: x86_64-unknown-linux-gnu
release: 1.90.0
LLVM version: 20.1.8
```

**Comment by TomFryersMidsummer** [comment]
Possibly related: #10579.
